### DISCUSSION TITLE
x-on:click doesn't work inside flickty on new versions

### DISCUSSION THREAD

**Discussion by mokhosh** [alpinejs/alpine#2329] [7. Bugs (archived)] [answered]
I just upgraded alpine to latest version and `x-on:click`s that i had inside a flickity slider have stopped working.
any ideas why this might be?
https://flickity.metafizzy.co/

**Comment by mokhosh** [comment]
Update: v3.0.8 is the latest version that works, and v3.0.9 is the first version that doesn't work.

**Comment by mokhosh** [comment]
This is really problematic because I'm using x-ref outside of x-init and that hasn't been fixed until several releases later. Now I just get an empty object when trying to reach x-ref in another event.

**Comment by mokhosh** [comment, marked as answer]
@SimoTod can you please see this?

**Reply by SimoTod** [reply]
Can you post a link to a simplified codepen or jfiddle replicating the issue? Trying to replicate without knowing exactly how to set up the script is time-consuming.

**Reply by mokhosh** [reply]
https://codepen.io/mokhosh/pen/vYJVNma

**Reply by SimoTod** [reply]
Thanks, I had a look.
So, first of all, there was a bug before 3.0.9 where children of elements removed from the DOM were not destroyed correctly (causing memory leaks and stuff like events, watcher and effects to keep firing after the element was removed).
3.0.9 fixed it which is why it was working for you (by chance, bear with me, the real issue is somewhere else).
1) Flickity removes those elements from the DOM (queueing an element removal in the alpine mutation observer) and re-adds those nodes wrapped in an additional div (which is the reason why alpine doesn't realise it's the same node. should it be more clever or is it okay to assume that the context has changed?)
2) A removal triggers `destroyTree` from the mutation observer (which removes the event listeners).
3) An addition triggers `initTree` from the mutation observer (which would re-add the event listeners).
4) While initTree runs straight away, destroyTree runs on nextTick so `destroyTree` always run after `initTree`. not sure if there is a reason for that. (Issue 1) -> https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/lifecycle.js#L16-L17
5) the code in the mutation observer actually tries to run additions first then removals (issue 2) -> https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/mutation.js#L163-L177
^ In order to fix them, we will need input from @calebporzio because there may be a reason why things are like that and it may be not trivial to change them.

AT that point you would end up with something like: https://codepen.io/SimoTod/pen/ZEJqmaW

**Reply by mokhosh** [reply]
This is amazing @SimoTod. Do you think I can use this custom alpine build in my production until @calebporzio gives his input on the issue?

**Reply by SimoTod** [reply]
all tests pass. It's fine but it's quite a low-level thing so I'm not 100% sure but I think it's fine (As a separate note, it seems to fix #2332 as well).

**Reply by SimoTod** [reply]
to help the discussion, Thes are the required changes: https://github.com/alpinejs/alpine/compare/main...SimoTod:bug/issue-with-mutations-and-third-party-libs?expand=1

**Reply by mokhosh** [reply]
These changes seem to work perfectly. Do you think we should make a PR, or do you think @calebporzio will see this and join the discussion?

**Reply by mokhosh** [reply]
@SimoTod this didn't make it to 3.5.2 🤦‍♂️  Do you think a PR would increase the chances of this being noticed sooner?

**Reply by SimoTod** [reply]
@mokhosh there was already a PR involving the mutation observer for this release so I didn't want to add this one at the same time. I'll raise a PR in the next few days after a i checked a couple of things.

**Reply by SimoTod** [reply]
@mokhosh https://github.com/alpinejs/alpine/pull/2399

**Comment by mokhosh** [comment]
I don't know if this is related but I get like 200+ of this error from Alpine dev tools:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'id')
```

**Reply by SimoTod** [reply]
I think that one is an error in Alpine dev tools.
